File: packages/evidence/src/digest.ts
import { createHash, timingSafeEqual } from 'node:crypto';
import type { Digest } from '@l9/assurance-contracts';
import { canonicalJson } from './canonical.js';

export function sha256Bytes(bytes: string | Uint8Array): Digest {
  return { algorithm: 'sha256', value: createHash('sha256').update(bytes).digest('hex') };
}
export function sha256Digest(value: unknown): Digest { return sha256Bytes(canonicalJson(value)); }
export function verifyDigest(value: unknown, digest: Digest): boolean {
  if (digest.algorithm !== 'sha256' || !/^[a-f0-9]{64}$/.test(digest.value)) return false;
  const actual = sha256Digest(value).value;
  return timingSafeEqual(new TextEncoder().encode(actual), new TextEncoder().encode(digest.value));
}

File: src/l9_assurance/evidence/digest.py
from __future__ import annotations

import hashlib
import hmac
from typing import Any

from .canonical import canonical_json


def sha256_bytes(data: str | bytes | bytearray | memoryview) -> dict[str, str]:
    payload = data.encode("utf-8") if isinstance(data, str) else bytes(data)
    return {"algorithm": "sha256", "value": hashlib.sha256(payload).hexdigest()}


def sha256_digest(value: Any) -> dict[str, str]:
    return sha256_bytes(canonical_json(value))


def verify_digest(value: Any, digest: object) -> bool:
    if not isinstance(digest, dict):
        return False
    if digest.get("algorithm") != "sha256":
        return False
    expected = digest.get("value")
    if not isinstance(expected, str) or len(expected) != 64:
        return False
    actual = sha256_digest(value)["value"]
    return hmac.compare_digest(actual, expected)
